registry/client.go

```diff
@@ -0,0 +1,278 @@
+
+const (
+	NoTimeout TimeoutType = iota
+	ReceiveTimeout
+	ConnectTimeout
+)
+
+// Client should handle http requests for all registry sessions.
+type Client struct {
+	httpClient *http.Client
+	isSecure   bool
+}
+
+// Do handles making an http request and ensures that
+// Auth headers are never sent over an insecure channel.
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	authHeader := req.Header.Get("Authorization")
+	if authHeader != "" && !(c.isSecure && req.URL.Scheme == "https") {
+		return nil, fmt.Errorf("cannot send auth credentials over insecure channel %q: %s", authHeader, req.URL)
+	}
+
+	return c.httpClient.Do(req)
+}
+
+// NewClient prepares and returns an HTTP Client to use with this endpoint.
+func (e *Endpoint) NewClient(jar http.CookieJar, timeout TimeoutType) *Client {
+	return newClient(jar, timeout, e.IsSecure)
+}
+
+func newClient(jar http.CookieJar, timeout TimeoutType, secure bool) *Client {
+	certLock.RLock()
+	tlsConfig := tls.Config{
+		RootCAs: caPool,
+		// Avoid fallback to SSL protocols < TLS1.0
+		MinVersion:   tls.VersionTLS10,
+		Certificates: clientCerts,
+	}
+	certLock.RUnlock()
+
+	if !secure {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	httpTransport := &http.Transport{
+		DisableKeepAlives: true,
+		Proxy:             http.ProxyFromEnvironment,
+		TLSClientConfig:   &tlsConfig,
+	}
+
+	switch timeout {
+	case ConnectTimeout:
+		httpTransport.Dial = func(proto string, addr string) (net.Conn, error) {
+			// Set the connect timeout to 5 seconds
+			d := net.Dialer{Timeout: 5 * time.Second, DualStack: true}
+
+			conn, err := d.Dial(proto, addr)
+			if err != nil {
+				return nil, err
+			}
+			// Set the recv timeout to 10 seconds
+			conn.SetDeadline(time.Now().Add(10 * time.Second))
+			return conn, nil
+		}
+	case ReceiveTimeout:
+		httpTransport.Dial = func(proto string, addr string) (net.Conn, error) {
+			d := net.Dialer{DualStack: true}
+
+			conn, err := d.Dial(proto, addr)
+			if err != nil {
+				return nil, err
+			}
+			conn = utils.NewTimeoutConn(conn, 1*time.Minute)
+			return conn, nil
+		}
+	}
+
+	return &Client{
+		httpClient: &http.Client{
+			Transport:     httpTransport,
+			CheckRedirect: AddRequiredHeadersToRedirectedRequests,
+			Jar:           jar,
+		},
+		isSecure: secure,
+	}
+}
+
+func getCertDirs() (certDirs []string) {
+	if err := os.MkdirAll(certsDirname, os.FileMode(0755)); err != nil {
+		log.Fatalf("unable to make docker registry certs directory %s: %s", certsDirname, err)
+	}
+
+	certDirs = []string{certsDirname}
+
+	// List directory entries to find all immediate subdirectories.
+	dirEntries, err := ioutil.ReadDir(certsDirname)
+	if err != nil {
+		log.Fatalf("unable to read certificate directory %s: %s", certsDirname, err)
+	}
+
+	for _, entry := range dirEntries {
+		if entry.IsDir() {
+			certDirs = append(certDirs, path.Join(certsDirname, entry.Name()))
+		}
+	}
+
+	return
+}
+
+func watchCertDirs() (certDirs []string) {
```

Comment:
Shouldn't the daemon require a restart or hup signal to reload certificates? It seems like a security hole to have this changing whenever the file system changes.